### PR TITLE
[distributed] Skip ProcessGroupNCCLTest if CUDA is not available

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28393 [distributed] Skip ProcessGroupNCCLTest if CUDA is not available**

We should skip this test if CUDA is not available and alert the user.
Previously, if this test was ran on cpu it would fail with:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  cuda runtime error (3) : This binary is linked with CUDA lazy stubs and underlying .so files were not loaded. CUDA functionality is disabled. Set env variable CUDA_LAZY_DEBUG to get messages during startup
```

Open question: should this file eventually be refactored to use `gtest` like we do for most other C++ tests?

Differential Revision: [D18054369](https://our.internmc.facebook.com/intern/diff/D18054369/)

### CODE REVIEW

**Review by mrshenli — APPROVED** [review]
Stamped. Besides adding this, I would assume there is some config to tell the internal test infra to only run this on GPU server? 

cc @yf225

**Comment by rohan-varma** [comment]
@pietern Would be great if you could take a look at this too

File: torch/lib/c10d/test/ProcessGroupNCCLTest.cpp [hunk 1]
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
 }
 
 int main(int argc, char** argv) {
+  if (!(c10::cuda::device_count() > 0)) {
```
Comment (pritamdamania87) [inline torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: nit: `if (!c10::cuda::is_available())`

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]

File: torch/lib/c10d/test/ProcessGroupNCCLTest.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: @pritamdamania87 this file doesn't use gtest right now, it should probably be rewritten to use gtest if we want to do this. For now, I've just changed this to `if (!c10::cuda::is_available())`. I've created a gh issue to track this: https://github.com/pytorch/pytorch/issues/28410

[the remark this answers is not in the document]
Comment (pietern) [reply torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: `GTEST_SKIP` is not available in the googletest version we use (1.8 AFAIK). Last time I checked it is only available on master (unreleased). We use googletest from a submodule so we could bump it to bring it in.

File: torch/lib/c10d/test/ProcessGroupNCCLTest.cpp [hunk 2]
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
 }
 
 int main(int argc, char** argv) {
+  if (!at::cuda::is_available()) {
+    // skip the test
```
Comment (pietern) [inline torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: Comment doesn't add anything and can go IMO.

File: torch/lib/c10d/test/ProcessGroupNCCLTest.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: I bumped the googletest to 1.10.0 which has `GTEST_SKIP`. The PR is here: https://github.com/pytorch/pytorch/pull/28432

File: torch/lib/c10d/test/ProcessGroupNCCLTest.cpp [hunk 3]
```diff
@@ -414,6 +414,11 @@ void testReduceScatter(const std::string& path, int rank, int size) {
...
+    std::cout << "CUDA not available, skipping test" << std::endl;
```
Comment (pritamdamania87) [inline torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: Is it possible to use glog here?

**Review by pritamdamania87 — APPROVED** [review]

File: torch/lib/c10d/test/ProcessGroupNCCLTest.cpp (hunk 3 above)
Comment (rohan-varma) [reply torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: It doesn't seem like this is used, but looks like we can do `LOG(INFO) << message` which seems to be the pattern in the codebase.

Comment (rohan-varma) [reply torch/lib/c10d/test/ProcessGroupNCCLTest.cpp]: Changed it to use `LOG(INFO) << ....`

**Pull request closed without merging**